### PR TITLE
Development branch

### PR DESCRIPTION
**Pull request by vvkuzmych**

### CODE REVIEW

File: lib/factory.rb [hunk 1]
```diff
@@ -14,3 +18,82 @@
...
+            instance_variable_set("@#{args[index]}", variables[index])
+          end
+        end
+
+        define_method :each do |&block|
+          values.each(&block)
+        end
+
+        define_method :each_pair do |&pair|
+          to_h.each_pair(&pair)
```
Comment (yaorlov) [inline lib/factory.rb]: let's implement it with `each` instead of `each_pair`

File: lib/factory.rb [hunk 2]
```diff
@@ -14,3 +18,82 @@
...
+        end
+
+        define_method :dig do |*args|
+          args.inject(self) { |values, args| values[args] if values }
+        end
+
+        define_method :to_h do
```
Comment (yuriy1991) [inline lib/factory.rb]: this method should be private

File: lib/factory.rb [hunk 3]
```diff
@@ -14,3 +18,82 @@
...
+
+        define_method :eql? do |other|
+          self.class == other.class && values == other.values
+        end
+
+        define_method(:to_a) do
+          instance_variables.map { |argument| instance_variable_get(argument) }
+        end
+
+        define_method(:values_at) do |*index|
```
Comment (yaorlov) [inline lib/factory.rb:73]: this can be implemented with `select` in one line

File: lib/factory.rb [hunk 4]
```diff
@@ -14,3 +18,82 @@
...
+            instance_variable_get values
+          end
+        end
+
+        define_method :[] do |argument|
+          argument.is_a?(Integer) ? values[argument] : instance_variable_get("@#{argument}")
+        end
+
+        define_method :[]= do |argument, value|
+          instance_variable_set("@#{argument}", value)
```
Comment (yaorlov) [inline lib/factory.rb]: what if an `argument` is an Integer?

File: .rubocop.yml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  Enabled: false
+
+Lint/ShadowingOuterLocalVariable:
+  Enabled: false
+
+Metrics/CyclomaticComplexity:
+  Max: 8
+
+Style/IfUnlessModifier:
+  Enabled: false
```
Comment (yaorlov) [inline .rubocop.yml]: new line

File: lib/factory.rb (hunk 3 above)
Comment (vvkuzmych) [reply lib/factory.rb:73]: done

File: lib/factory.rb [hunk 6]
```diff
@@ -43,7 +43,7 @@ def create_new_class(*args, &block)
         end
 
         define_method :each_pair do |&pair|
```
Comment (yaorlov) [inline lib/factory.rb]: please rename `pair` argument to `block`

File: lib/factory.rb [hunk 7]
```diff
@@ -75,17 +75,15 @@ def create_new_class(*args, &block)
...
+          values.select { |value| index.include?(values.index(value)) }
         end
 
         define_method :[] do |argument|
           argument.is_a?(Integer) ? values[argument] : instance_variable_get("@#{argument}")
         end
 
         define_method :[]= do |argument, value|
-          instance_variable_set("@#{argument}", value)
+          argument.is_a?(Integer) ? values[argument] : instance_variable_set("@#{argument}", value)
```
Comment (yaorlov) [inline lib/factory.rb]: how the value is assigned when an argument is an Integer?

File: lib/factory.rb [hunk 8]
```diff
@@ -91,6 +87,12 @@ def create_new_class(*args, &block)
         alias_method :==, :eql?
 
         class_eval(&block) if block_given?
+
+        private
```
Comment (yaorlov) [inline lib/factory.rb]: please remove `private`

File: lib/factory.rb [hunk 9]
```diff
@@ -14,3 +18,80 @@
...
+        define_method(:values_at) do |*index|
+          values.select { |value| index.include?(values.index(value)) }
+        end
+
+        define_method :[] do |argument|
+          argument.is_a?(Integer) ? values[argument] : instance_variable_get("@#{argument}")
+        end
+
+        define_method :[]= do |argument, value|
+          instance_variable_set argument.is_a?(Integer) ? instance_variables[argument] : "@#{argument}", value
```
Comment (yaorlov) [inline lib/factory.rb]: please split `instance_variable_set` and ternary operator into 2 lines

**Review by yaorlov — APPROVED** [review]

**Pull request merged**